### PR TITLE
Scheduling API

### LINKED ISSUE #127: RFC: Scheduling API
This issue outlines a proposed scheduling API for taco. The API consists of a taco-specific schedule primitive called operator splitting, and a number of loop optimization primitives.

The API applies schedule operations to the index variables of expressions, and are exposed as methods on `Expr` (operator split) and `Tensor` (loop optimizations) in the `index_expression` module.

# Operator Split

The operator split optimization splits an index variable with respect to an index expression. A workspace tensor is introduced as the result of the expression of the left index variable and as an operand of the right index variable.

Operator split can be called on an `Expr` to split it:

```c++
IndexVar i, ib, ic;
IndexExpr mul = b(i) * c(i);
mul.splitOperator(i, ib, ic);
```

Assignments may also be split, but this is done automatically by the compiler when it is necessary.

# Loop Optimizations

The loop optimization primitives lets the user express traditional loop optimizations. For confirmity we adapt the naming and method declarations of [Halide](http://halide-lang.org/docs/tutorial_2lesson_05_scheduling_1_8cpp-example.html). The semantics, however, changes somewhat because index expressions are not the same as halide functions and because taco generates sparse loops.

#### Reorder

Reorder changes the default ordering, which is deduced from the format layout directions (e.g. row-major vs. column-major). Reordering may cause edges in iteration graphs to point upwards, which is fine for dense formats, but otherwise (e.g. sparse formats) requires scanning.

```c++
IndexVar i, j;
A(i,j) = B(i,j) + C(i,j);
A.reorder(j, i);
```

#### Split

Split splits (strip-mines) a loop into an outer loop and an inner loop of fixed size.  The benefit is that the inner loop is guaranteed to be a fixed size, and so can be vectorized, etc.  This operation has to be made to work with sparse variable-sized formats, but I think it can work.

```c++
IndexVar i, iOuter, iInner;
a(i) = b(i) + c(i);
a.split(i, iOuter, iInner, 4);
```

#### Tile

Shortcut for splitting and reordering two index variables.

```c++
IndexVar i, j, iOuter, iInner, jOuter, jInner;
A(i,j) = B(i,j) + C(i,j);
A.tile(i, j, iOuter, jOuter, iInner, jInner, 4, 4);
```

#### Fuse

Fuses two loops, which enables more efficient non-nested parallelization (especially important on GPUs). It can only be possible to fuse fixed-width loops.

- [x] Do we need to support `fuse`?

```c++
IndexVar i, j;
A(i,j) = B(i,j) + C(i,j);
A.fuse(i,j);
```

#### Unroll

Unroll a fixed-size loop. Can either be called on a fixed-size loop, or supplied with a split factor. 

```c++
IndexVar i, j;
A(i,j) = B(i,j) + C(i,j);
A.unroll(i, 4);
```

#### Vectorize

Vectorize a fixed-size loop. Can either be called on a fixed-size loop, or supplied with a split factor. This loop must meet the parallelization preconditions.

```c++
IndexVar i, j;
A(i,j) = B(i,j) + C(i,j);
A.vectorize(i, 4);
```

- [ ] Can this be used with split and reorder to implement lane vectorization?

#### Parallelize

Parallelize a loop. This loop must meet the parallelization preconditions.

```c++
IndexVar i, j;
A(i,j) = B(i,j) + C(i,j);
A.parallel(i);
```

### LINKED ISSUE #129: GPU Code Generation
We need to generate GPU kernels. As a first step we're planning to generate CUDA code and compile it with our current system-compiler scheme (Nvidia has graciously donated GPUs we'll work with).

The target for GPU support in taco is mid April.

Open questions:
- [ ] How should we design the device API to move tensors to and from GPU memories?
- [ ] What formats will work best on GPUs?
- [ ] What scheduling constructs are needed on GPUs (see #127)?
- [ ] What does assembly code look like on GPUs?
- [ ] Do users need other backends than CUDA (e.g., OpenCL, OpenACC, or Metal)?

### PR DESCRIPTION
**Pull request by RSenApps**
Adds a scheduling language to TACO that allows applying optimizations to sparse tensor algebra kernels. The implemented transformations pos, coord, reorder, fuse, split, divide, parallelize, bound, and unroll allow for generating sparse tensor algebra kernels for CPU and GPU that match the performance of hand-tuned libraries. For more details of how to use these optimizations see http://www.rsenapps.com/senanayake-thesis.pdf (especially Section 7, which details the API). High-performance schedule examples are provided in tests-scheduling-eval.cpp. Closes #129, closes #127

### CODE REVIEW

File: include/taco/index_notation/provenance_graph.h [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+  virtual std::vector<ir::Expr> computeRelativeBound(std::set<IndexVar> definedVars, std::map<IndexVar, std::vector<ir::Expr>> computedBounds, std::map<IndexVar, ir::Expr> variableExprs, Iterators iterators, ProvenanceGraph provGraph) const;
+
+  /// Determine the iteration bounds for a newly derived index variable given the iteration bounds of the parents
+  virtual std::vector<ir::Expr> deriveIterBounds(IndexVar indexVar, std::map<IndexVar, std::vector<ir::Expr>> parentIterBounds, std::map<IndexVar, std::vector<ir::Expr>> parentCoordBounds,
+                                                 std::map<taco::IndexVar, taco::ir::Expr> variableNames,
+                                                 Iterators iterators, ProvenanceGraph provGraph) const;
+
+  /// Recover a parent index variable expression as a function of the children index variables
+  virtual ir::Expr recoverVariable(IndexVar indexVar, std::map<IndexVar, ir::Expr> variableNames, Iterators iterators, std::map<IndexVar, std::vector<ir::Expr>> parentIterBounds, std::map<IndexVar, std::vector<ir::Expr>> parentCoordBounds, ProvenanceGraph provGraph) const;
```
Comment (amaleewilson) [inline include/taco/index_notation/provenance_graph.h:69]: Some of these lines are very long. Would be nice to setup git clang format.

Comment (RSenApps) [reply include/taco/index_notation/provenance_graph.h:69]: I agree, I wish I was doing this throughout. I can run clang-format on this entire pull request, but am also nervous to make large changes now that Rawn and Tony have been working off this branch for several months. I will talk to them about it thanks.

File: src/codegen/codegen_cuda.cpp [hunk 2]
```diff
@@ -63,7 +63,100 @@ const string gpuAssertMacro =
   "    fprintf(stderr,\"GPUassert: %s %s %d\\n\", cudaGetErrorString(code), file, line);\n"
   "    if (abort) exit(code);\n"
   "  }\n"
+  "}\n"
+  "__device__ __host__ int taco_binarySearchAfter(int *array, int arrayStart, int arrayEnd, int target) {\n"
```
[the remark this answers is not in the document]
Comment (RSenApps) [reply src/codegen/codegen_cuda.cpp:67]: Ideally we would just have these library functions in a separate header file (this is what we do in the evaluation framework repo). The existing codebase outputted strings like this so I kept the existing style, but I agree that it is not sustainable as the number of library functions grow.

File: include/taco/index_notation/provenance_graph.h (hunk 1 above)
Comment (amaleewilson) [reply include/taco/index_notation/provenance_graph.h:69]: Yeah, I definitely didn't mean to suggest it should be done now. Maybe after the merge there could be a refactor commit on master along with some formatting guidelines? Clang format is pretty easy to setup, but I totally understand not wanting to throw in a whole bunch of changes while people are working.

Comment (RSenApps) [reply include/taco/index_notation/provenance_graph.h:69]: Ok I agree that would be a good way to go about it.

**Pull request merged**